### PR TITLE
DOC: NAV-134 - Service implementation

### PR DESCRIPTION
**Pull request by munterfi**

### CODE REVIEW

File: Writerside/topics/implementation/public-transit-service.md [hunk 1]
```diff
@@ -1,9 +1,166 @@
...
+package raptor.router {
+    class RaptorRouter {
+    }
+}
+
+package service.gtfs.raptor.convert {
+    class GtfsToRaptorConverter {
+    }
+    class GtfsRoutePartitioner {
+    }
```
Comment (clukas1) [inline Writerside/topics/implementation/public-transit-service.md:58]: maybe add DayTripMaskProvider, although not explained yet.

File: Writerside/topics/implementation/public-transit-service.md [hunk 2]
```diff
@@ -1,9 +1,166 @@
...
+}
+
+PublicTransitService <|... GtfsRaptorService: <<implements>>
+PublicTransitService --|> ScheduleInformationService: <<extends>>
+PublicTransitService --|> ConnectionRoutingService: <<extends>>
+
+GtfsRaptorService --> GtfsToRaptorConverter : uses
+GtfsRaptorService *- WalkCalculator : has
+GtfsRaptorService *- GtfsSchedule : has
```
Comment (clukas1) [inline Writerside/topics/implementation/public-transit-service.md:67]: Just wondering if we should also add that the GtfsRaptorService creates the Raptor Instance in it's constructor from the Gtfs Schedule

File: Writerside/topics/implementation/public-transit-service.md [hunk 3]
```diff
@@ -1,9 +1,166 @@
...
+The GTFS model and RAPTOR algorithm have differing perspectives on routes. In GTFS, a "route" refers to a user-friendly
+view of trips that may share a common line number or route name, even if the stop sequences differ slightly. For
+instance, a regional train might have trips that skip certain stops on later trips while still being part of the same
+route from a user's point of view.
+
+However, RAPTOR requires that all trips associated with a route must have the exact same stop sequence. This discrepancy
+is resolved by the `GtfsToRaptorConverter` by applying the `GtfsRoutePartitioner`, which splits GTFS routes into "
+sub-routes". These sub-routes group trips that share identical stop sequences, ensuring compatibility with RAPTOR’s
+routing model. During post-processing, when the routing results are returned to the user, the sub-routes are converted
+back into the original GTFS routes to maintain user-friendly responses (TODO: Where is this happening? in the RAPTOR or
```
Comment (clukas1) [inline Writerside/topics/implementation/public-transit-service.md:130]: This is done in the Service this happens in the TypeMapper:

```
    public static Connection map(ch.naviqore.raptor.Connection connection, @Nullable Walk firstMile,
                                 @Nullable Walk lastMile, GtfsSchedule schedule) {
        List<Leg> legs = new ArrayList<>();

        if (firstMile != null) {
            legs.addFirst(firstMile);
        }

        for (ch.naviqore.raptor.Leg leg : connection.getLegs()) {
            legs.add(map(leg, schedule));
        }

        if (lastMile != null) {
            legs.addLast(lastMile);
        }

        return new GtfsRaptorConnection(legs);
    }

    public static Leg map(ch.naviqore.raptor.Leg leg, GtfsSchedule schedule) {
        int duration = (int) Duration.between(leg.getDepartureTime(), leg.getArrivalTime()).toSeconds();
        Stop sourceStop = map(schedule.getStops().get(leg.getFromStopId()));
        Stop targetStop = map(schedule.getStops().get(leg.getToStopId()));
        int distance = (int) Math.round(sourceStop.getLocation().distanceTo(targetStop.getLocation()));

        return switch (leg.getType()) {
            case WALK_TRANSFER ->
                    new GtfsRaptorTransfer(distance, duration, leg.getDepartureTime(), leg.getArrivalTime(), sourceStop,
                            targetStop);
            case ROUTE -> createPublicTransitLeg(leg, schedule, distance);
        };
    }
```

File: Writerside/topics/implementation/public-transit-service.md [hunk 4]
```diff
@@ -1,9 +1,166 @@
...
+mile), which represent the walk from the start coordinate to the selected origin stop and from the destination stop to
+the final coordinate.
+
+In summary, the `ConnectionRoutingService` performs pre-processing to identify nearby stops, routes the request using
+the RAPTOR engine, and post-processes the results by adding walking segments and re-aggregating sub-routes to provide
+user-friendly routing information.
+
+**Note:** Since we also support routing based on the latest departure, which involves routing backwards in time, we use
+the terms "source" and "target" stop in the code. However, in this explanation, we use the more traditional terms "
+origin" and "destination" for clarity.
```
Comment (clukas1) [inline Writerside/topics/implementation/public-transit-service.md:166]: unnecessary white space between " and origin

File: Writerside/topics/implementation/public-transit-service.md [hunk 5]
```diff
@@ -1,9 +1,166 @@
...
+
+The walking distances between stops are calculated using a `WalkCalculator`. This implementation uses a **beeline
+distance factor**, typically set to 1.3 based on literature [13, p. 43], to account for real-world walking conditions
+such as obstacles, detours, or building access routes. The calculator multiplies the straight-line distance by this
+factor to approximate a more realistic walking time. The `WalkCalculator` interface allows for easy replacement with
+more advanced footpath routing solutions, such as A* or A+ Landmark algorithms, if needed.
+
+### Handling Coordinates
+
+In routing requests, users can specify coordinates (latitude/longitude) or stops as the origin or destination. When
```
Comment (clukas1) [inline Writerside/topics/implementation/public-transit-service.md:149]: This is actually also a point we can add to the extended raptor, that we modified / extended the algorithm to allow setting multiple source stops (with different source times) and target multiple target stops (with different handicaps)

File: Writerside/topics/implementation/public-transit-service.md (hunk 1 above)
Comment (munterfi) [reply Writerside/topics/implementation/public-transit-service.md]: https://github.com/naviqore/documentation/pull/30/commits/1c0f164446f85efd376c2b84dde565013bbe611b

File: Writerside/topics/implementation/public-transit-service.md (hunk 2 above)
Comment (munterfi) [reply Writerside/topics/implementation/public-transit-service.md:70]: Yes you are right, we directly rely on the RaptorRouter and not on the RaptorAlgorithm interface in the GtfsRaptorService. Fixed in: https://github.com/naviqore/documentation/pull/30/commits/fc8274a2f8c22b857bcb42ff72ee988af97f935f

File: Writerside/topics/implementation/public-transit-service.md (hunk 3 above)
Comment (munterfi) [reply Writerside/topics/implementation/public-transit-service.md]: https://github.com/naviqore/documentation/pull/30/commits/5f58b719276cad77e0558a29be950566a692c7c5

File: Writerside/topics/implementation/public-transit-service.md (hunk 4 above)
Comment (munterfi) [reply Writerside/topics/implementation/public-transit-service.md]: https://github.com/naviqore/documentation/pull/30/commits/5a613a8cbbb969087380d864b80035f745544b9d

File: Writerside/topics/implementation/public-transit-service.md (hunk 5 above)
Comment (munterfi) [reply Writerside/topics/implementation/public-transit-service.md:153]: Good point, I added a TODO.

**Comment by munterfi** [comment]
@clukas1: Now the TripMaskProvider and StopTimesProvider is duplicated in the class diagrams of the extended RAPTOR and the public transit service class diagram. Maybe we delete it from the service class diagram?

**Comment by munterfi** [comment]
Hmm ok, this is the design for the old implementation... so I think it is good to keep it in both class diagrams.

For completeness we could add this updated UML diagram for the new design:

```plantuml
@startuml
set namespaceSeparator none

package gtfs.schedule {
    class GtfsSchedule
}

package raptor.router {
  class RaptorRouter {
    + routeEarliestArrival(...): List<Connection>
  }

  class StopTimeProvider {
  + getStopTimesForDate(LocalDate date, ...): int[]
  }
    
  interface RaptorTripMaskProvider {
    + getTripMasksForDay(): DayTripMask
  }
  
  class DayTripMask {
    - tripMask: Map<String, boolean[]>
  }
}

package service.gtfs.raptor.convert {
  class GtfsTripMaskProvider
}

GtfsTripMaskProvider .|> RaptorTripMaskProvider: "<<implements>>"
RaptorRouter *- StopTimeProvider : has
StopTimeProvider *-- RaptorTripMaskProvider : has
RaptorTripMaskProvider --> DayTripMask: "provides"
StopTimeProvider --> DayTripMask: "uses"
GtfsTripMaskProvider o-- GtfsSchedule: has

@enduml
```

**Review by clukas1 — APPROVED** [review]
Looks good from my side

**Pull request merged**